### DISCUSSION TITLE
Record with keyword used with inheritance seems to replace the reference with it's own

### DISCUSSION THREAD

**Discussion by bogdan-patraucean** [dotnet/csharplang#8001] [Q&A]
I think I discovered an issue with how records work in combination with inheritance. Updating an object and making a copy of it using the `with` keyword will replace the old reference of the base class with it's own (the derived class).

To showcase this we need 3 classes, and as example ([fiddle here](https://dotnetfiddle.net/p5bANi)) I chose state management:
- an abstract record `GenericState` class to define what a state is
- a `State` class where we will store all our data
- a `Substate` class that's part of the `State` (which is the root/parent)

```
public abstract record class GenericState
{
  public GenericState? Parent {get; init; }
  
  public abstract string Name {get; }
  
  public GenericState(GenericState? parent)
  {
    this.Parent = parent;
  }
}

public record class Substate: GenericState
{
  public override string Name => "Child";
  
  public int Property {get;set;}
  
  public Substate(State parent): base(parent) { }
}

public record class State: GenericState
{
  public override string Name => "Parent";
  
  public State(GenericState parent): base(parent)
  {
    child = new(this);
  }
  
  private Substate child;
  public Substate Child
  {
    get => child;
    init
    {
      if (child == value)
      {
        return;
      }
      
      child = value;
    }
  }
}
```

And here is what I do, I create a state with a child, and I specify the parent to the child so it always knows who the parent is. The state will not have any parent so I initialize it with null since it's the root. When I try to update the state by updating the child's property to 4 the reference to the parent will be replaced with the reference of the child.

So, the first log to console will display `Parent`, while the second will display `Child`, tho, `with` should copy everything as is.

```
var state= new State(null);
var newState = state with { Child = state.Child with { Property = 4 } };

Console.WriteLine(state.Child.Parent.Name);
Console.WriteLine(newState.Child.Parent.Name);
```

Am I doing something wrong? For me this seems like an issue, as I would expect the reference to be kept, not replaced.

## Note

The same example without the abstract record class works as expected ([fiddle here](https://dotnetfiddle.net/FM7NCm)).

**Comment by HaloFour** [comment]
You're using `with` on the `Child`, so it is expected to be replaced with the new modified instance.  Is that not what you're seeing?

**Reply by bogdan-patraucean** [reply]
The Child initially has a reference to the parent, why is it not kept? it should be kept, not replaced with itself. I believe that has to do with how records work under the hood.

**Reply by HaloFour** [reply]
I think the issue you're running into is that the synthesized `Clone` method used by `with` calls a copy constructor which accepts the instance of the class.  That constructor is what actually copies the state over to the new instance, but you've replaced it with one that does something different.

Normally, when you declare a record like this:

```cs
public record State { }
```

It synthesizes the following to support `with`:

```cs
public class State {
    [CompilerGenerated]
    protected State(State original) {
        // copy state here
    }

    [CompilerGenerated]
    public State <Clone>$() {
        return new State(this);
    }
}
```

So when you added the custom constructor to `GenericState`, the compiler thinks you're providing a custom copy constructor, which is getting invoked with the new state instance.

**Reply by bogdan-patraucean** [reply]
Maybe, the same example without the abstract class works.

**Reply by 333fred** [reply]
Halo is correct. When you manually override the copy constructor like this, you get what you implemented.

**Reply by HaloFour** [reply]
I can see where it would be confusing, especially if you're not aware that records rely on a copy constructor or that you can override it.  Positional records do appear to forbid declaring a primary constructor that only accepts the record type as the only argument, but there's no guardrails for non-positional records.  But given it's something that some developers will want to do intentionally it certainly doesn't feel like it should emit a warning either.

**Reply by bogdan-patraucean** [reply]
And how could I avoid this and obtain the behavior I need? Not using inheritance works in the same way.

**Reply by HaloFour** [reply]
I think you'd might need to use the `init` property to set the `Parent` state.  If that doesn't enforce the contract you want you might need to use a static factory method to create new instances of the state.  Or if there is other required state you can include that parameter in the constructor so it's not confused with the copy constructor.

**Reply by huoyaoyuan** [reply]
> When you manually override the copy constructor like this, you get what you implemented.

I'd like to see some IDE support that tells you this will be used as copy constructor. Users not familiar with records may not even be aware to find the doc about this.

**Comment by bogdan-patraucean** [comment]
I confirm that adding another parameter to differentiate itself from the copy constructor of the record achieves what I want. Thanks @HaloFour for pointing this out, an updated fiddle can be found [here](https://dotnetfiddle.net/Js1l0d). Tho, a question still remains: do we want the copy constructor to be overriden so easely and by mistake?

Is there nothing we can do to prevent that?

**Reply by Richiban** [reply]
It may be because your example has been reduced, but I don't see why you've written your own constructors for these types at all. Did the primary constructors not give you some functionality you wanted?

**Reply by bogdan-patraucean** [reply]
I'm not sure I follow you. Primary constructors only allow public properties, they should not be used in combination with abstraction just for the sake of it. There are some limits and using them doesn't really link to my issue which is with records.

**Reply by HaloFour** [reply]
I guess the question is that the example in your OP could have been written to use primary constructors.  There may be other aspects that were reduced out that would have precluded this.  However, attempting to use primary constructors would have run into the error that the primary constructor and copy constructor would have collided.
